### PR TITLE
Box_intersection_d: Make variables thread local

### PR DESCRIPTION
**Pull request by afabri**
Solves #1388

### CODE REVIEW

File: Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h [hunk 1]
```diff
@@ -36,14 +38,16 @@ namespace Box_intersection_d {
 template <class Dummy>
 struct Unique_numbers {
     typedef std::size_t ID;
-    Unique_numbers() : i(n++) {}
+    Unique_numbers() {
+      CGAL_STATIC_THREAD_LOCAL_VARIABLE(std::size_t, n, 0);
+      i = n++;
+    }
```
Comment (sloriot) [inline Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h]: Similarly here, I think it should be a `tbb::atomic<std::size_t>`

cc @afabri @lrineau

File: Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h [hunk 2]
```diff
@@ -36,14 +38,16 @@ namespace Box_intersection_d {
 template <class Dummy>
 struct Unique_numbers {
     typedef std::size_t ID;
-    Unique_numbers() : i(n++) {}
+    Unique_numbers() {
+      CGAL::cpp11::atomic<std::size_t>(n); // initialized to 0
+      i = n++;
```
Comment (sloriot) [inline Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h]: I guess you're missing a static I guess

Comment (afabri) [inline Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h]: fixed

File: Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h [hunk 3]
```diff
@@ -36,14 +38,16 @@ namespace Box_intersection_d {
 template <class Dummy>
 struct Unique_numbers {
     typedef std::size_t ID;
-    Unique_numbers() : i(n++) {}
+    Unique_numbers() {
+      static CGAL::cpp11::atomic<std::size_t>(n); // initialized to 0
```
Comment (lrineau) [inline Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h]: Have you tried to compile that code?! There should be at least a warning.
Use:
```C++
static CGAL::cpp11::atomic<std::size_t> n; // initialized to 0
```

File: Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h [hunk 4]
```diff
@@ -275,9 +275,8 @@ split_points( RandomAccessIter begin, RandomAccessIter end,
 
 
 #if CGAL_BOX_INTERSECTION_DEBUG
- static int level = -1;
  #define CGAL_BOX_INTERSECTION_DUMP(msg) { \
-   for( unsigned int i = level; i; --i ) \
+   for( unsigned int i = -1; i; --i ) \
```
Comment (lrineau) [inline Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h]: This patch completely modifies the loop. As such the loop will loop from `-1`, that is `UINTMAX` as a `unsigned`, decreasing to to `0`, whereas before the patch it was looping from `level` to `0.

Comment (afabri) [inline Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h]: I had not really understood the code. The variable `Counter<int> bla(level);`  increases level each time the function is called, so the for loop starts with `level == 0`.

File: Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h [hunk 5]
```diff
@@ -36,14 +38,16 @@ namespace Box_intersection_d {
 template <class Dummy>
 struct Unique_numbers {
     typedef std::size_t ID;
-    Unique_numbers() : i(n++) {}
+    Unique_numbers() {
+      static CGAL::cpp11::atomic<std::size_t> n; // initialized to 0
+      i = n++;
+    }
```
Comment (lrineau) [inline Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h]: As `n` is now an `atomic<integral>`, [its post-increment `operator++ is equivalent to `n.fetch_add(1)`][*]. That synchronizes with other threads.

[*]: http://en.cppreference.com/w/cpp/atomic/atomic/operator_arith

Comment (lrineau) [inline Box_intersection_d/include/CGAL/Box_intersection_d/Box_d.h]: It means that those "unique IDs" will really be unique, even if there are multiple threads.

File: Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h [hunk 6]
```diff
@@ -339,6 +338,7 @@ void segment_tree( RandomAccessIter1 p_begin, RandomAccessIter1 p_end,
     const T sup = box_limits< T >::sup();
 
 #if CGAL_BOX_INTERSECTION_DEBUG
+  CGAL_STATIC_THREAD_LOCAL_VARIABLE(int, level, -1);
```
Comment (lrineau) [inline Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h]: `level` is used for debugging only. it being a thread-local variable is the right choice: anyway the `Box_intersection_d` is not thread-safe if the same data is concurrently used by multiple threads.

**Comment by lrineau** [comment]
There are errors with older platforms, like:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.10-Ic-3/Box_intersection_d/TestReport_lrineau_CentOS6.gz

`CGAL::cpp11::atomic<T>` will not compile
  - if `CGAL_NO_THREADS` is defined,
  - or C++11 cannot be used and Boost<1.53.

If `CGAL_NO_THREADS`, `CGAL::cpp11::atomic<T>` could just be equivalent to `T` and that would work.

But what can we do in both the compiler and Boost are too old?

**Pull request merged**